.github/workflows/ci.yml

```diff
@@ -778,34 +724,26 @@ jobs:
           # final artifact name can't be longer than 128 characters
           echo "ARTIFACT_NAME=${name:0:100}" >> $GITHUB_ENV
       - name: Upload test results
-        uses: actions/upload-artifact@v4
-        # Upload all test reports only on failure, because the artifacts are large
-        if: failure()
-        with:
-          name: result ${{ env.ARTIFACT_NAME }}
-          path: |
-            **/target/surefire-reports
-            **/target/checkstyle-*
-      - name: Upload test report
-        uses: actions/upload-artifact@v4
-        # Always upload the test report for the annotate.yml workflow,
-        # but only the single XML file to keep the artifact small
+        uses: ./.github/actions/process-test-results
         if: always()
         with:
-          # Name prefix is checked in the `Annotate checks` workflow
-          name: test report ${{ github.job }} (${{ env.ARTIFACT_NAME }})
-          path: |
-            **/surefire-reports/TEST-*.xml
-          retention-days: ${{ env.TEST_REPORT_RETENTION_DAYS }}
-      - name: Upload heap dump
-        uses: actions/upload-artifact@v4
-        if: failure() && env.SECRETS_PRESENT == '' && github.event_name == 'pull_request' && github.event.pull_request.head.repo.full_name != github.repository
-        with:
-          name: heap dump ${{ github.job }} (${{ env.ARTIFACT_NAME }})
-          if-no-files-found: 'ignore'
-          path: |
-            **/*.hprof
-          retention-days: ${{ env.HEAP_DUMP_RETENTION_DAYS }}
+          artifact-name: ${{ env.ARTIFACT_NAME }}
+          has-failed-tests: >-
+            ${{ steps.tests.outcome == 'failure'
+              || steps.tests-hdfs-isolated.outcome == 'failure'
+              || steps.tests-hdfs.outcome == 'failure'
+              || steps.tests-s3.outcome == 'failure'
+              || steps.tests-azure.outcome == 'failure'
+              || steps.tests-gcs.outcome == 'failure'
+              || steps.tests-delta.outcome == 'failure'
+              || steps.tests-memsql.outcome == 'failure'
+              || steps.tests-bq.outcome == 'failure'
+              || steps.tests-bq-ci.outcome == 'failure'
+              || steps.tests-bq-smoke.outcome == 'failure'
+              || steps.tests-iceberg.outcome == 'failure'
+              || steps.tests-redshift.outcome == 'failure'
+            }}
```

Comment:
How we gonna keep this up to date?
Maybe we can have a test that takes ci.yml file and asserts that (1) each test task has ID and (2) all are mentioned here?